server/src/test/java/org/elasticsearch/snapshots/mockstore/MockEventuallyConsistentRepository.java

```diff
@@ -0,0 +1,186 @@
+ *     <li>Blobs can become visible to list operations with a delay.</li>
+ * </ul>
+ */
+public class MockEventuallyConsistentRepository extends FsRepository {
+
+    private final DeterministicTaskQueue deterministicTaskQueue;
+
+    private final Context context;
+
+    public MockEventuallyConsistentRepository(RepositoryMetaData metadata, Environment environment,
+                          NamedXContentRegistry namedXContentRegistry, DeterministicTaskQueue deterministicTaskQueue, Context context) {
+        super(metadata, environment, namedXContentRegistry, deterministicTaskQueue.getThreadPool());
+        this.deterministicTaskQueue = deterministicTaskQueue;
+        this.context = context;
+    }
+
+    @Override
+    protected void assertSnapshotOrGenericThread() {
+        // eliminate thread name check as we create repo in the test thread
+    }
+
+    @Override
+    protected BlobStore createBlobStore() throws Exception {
+        return new MockBlobStore(super.createBlobStore());
+    }
+
+    /**
+     * Context that must be shared between all instances of {@link MockEventuallyConsistentRepository} in a test run.
+     */
+    public static final class Context {
+
+        /**
+         * Map of blob path to a tuple of cached non-existent blobs in them and a map of child blob name to {@link Runnable} that when
+         * executed will create the blob.
+         */
+        private final Map<BlobPath, Tuple<Set<String>, Map<String, Runnable>>> state = new HashMap<>();
+
+        public Tuple<Set<String>, Map<String, Runnable>> getState(BlobPath path) {
+            return state.computeIfAbsent(path, p -> new Tuple<>(new HashSet<>(), new HashMap<>()));
+        }
+    }
+
+    private class MockBlobStore extends BlobStoreWrapper {
+
+        MockBlobStore(BlobStore delegate) {
+            super(delegate);
+        }
+
+        @Override
+        public BlobContainer blobContainer(BlobPath path) {
+            return new MockBlobContainer(super.blobContainer(path), context.getState(path));
+        }
+
+        private class MockBlobContainer extends BlobContainerWrapper {
+
+            private final Set<String> cachedMisses;
+
+            private final Map<String, Runnable> pendingWrites;
+
+            MockBlobContainer(BlobContainer delegate, Tuple<Set<String>, Map<String, Runnable>> state) {
+                super(delegate);
+                cachedMisses = state.v1();
+                pendingWrites = state.v2();
+            }
+
+            @Override
+            public boolean blobExists(String blobName) {
+                ensureReadAfterWrite(blobName);
+                final boolean result = super.blobExists(blobName);
+                if (result == false) {
+                    cachedMisses.add(blobName);
+                }
+                return result;
+            }
+
+            @Override
+            public InputStream readBlob(String name) throws IOException {
+                ensureReadAfterWrite(name);
```

Comment:
I'm not sure, how we're simulating eventual consistency for non-first write operation here.
If there is pendingWrite, it seems that we're just executing it and perform read of the latest pendingWrite.

Reply:
That's a fair point, thanks for spotting this! Practically it doesn't change anything (fortunately) since the `index.latest` blob is the only blob we ever overwrite (and we don't use that blob in normal blob repository operation) but I'll fix this shortly :)